File: app/src/main/java/closer/vlllage/com/closer/handler/search/SearchGroupsAdapter.java
package closer.vlllage.com.closer.handler.search;

import android.support.annotation.DrawableRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.handler.data.AccountHandler;
import closer.vlllage.com.closer.handler.data.ApiHandler;
import closer.vlllage.com.closer.handler.event.EventDetailsHandler;
import closer.vlllage.com.closer.handler.group.GroupActionAdapter;
import closer.vlllage.com.closer.handler.group.GroupMemberHandler;
import closer.vlllage.com.closer.handler.helpers.ActivityHandler;
import closer.vlllage.com.closer.handler.helpers.ApplicationHandler;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.helpers.ImageHandler;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.pool.PoolRecyclerAdapter;
import closer.vlllage.com.closer.pool.TempPool;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.Event;
import closer.vlllage.com.closer.store.models.Event_;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.GroupAction;
import closer.vlllage.com.closer.store.models.GroupAction_;
import io.objectbox.android.AndroidScheduler;
import jp.wasabeef.picasso.transformations.BlurTransformation;

import static closer.vlllage.com.closer.pool.Pool.tempPool;

public class SearchGroupsAdapter extends PoolRecyclerAdapter<SearchGroupsAdapter.SearchGroupsViewHolder> {

    private String createPublicGroupName = null;
    private final List<Group> groups = new ArrayList<>();
    private OnGroupClickListener onGroupClickListener;
    private OnCreateGroupClickListener onCreateGroupClickListener;
    private String actionText;
    private @LayoutRes int layoutResId = R.layout.search_groups_item;
    private @DrawableRes int backgroundResId = R.drawable.clickable_light;
    private boolean isSmall;

    public SearchGroupsAdapter(PoolMember poolMember, OnGroupClickListener onGroupClickListener, OnCreateGroupClickListener onCreateGroupClickListener) {
        super(poolMember);
        this.onGroupClickListener = onGroupClickListener;
        this.onCreateGroupClickListener = onCreateGroupClickListener;
    }

    public SearchGroupsAdapter setLayoutResId(int layoutResId) {
        this.layoutResId = layoutResId;
        return this;
    }

    public SearchGroupsAdapter setBackgroundResId(int backgroundResId) {
        this.backgroundResId = backgroundResId;
        return this;
    }

    @NonNull
    @Override
    public SearchGroupsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new SearchGroupsViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(layoutResId, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull SearchGroupsViewHolder holder, int position) {
        if (position >= getItemCount() - getCreateGroupCount()) {
            holder.pool = tempPool();
            holder.action.setText($(ResourcesHandler.class).getResources().getString(R.string.create_group));
            holder.name.setText(createPublicGroupName);
            holder.about.setText($(ResourcesHandler.class).getResources().getString(R.string.add_new_public_group));
            holder.backgroundPhoto.setVisibility(View.GONE);
            holder.actionRecyclerView.setVisibility(View.GONE);
            holder.cardView.setOnClickListener(view -> {
                if (onCreateGroupClickListener != null) {
                    onCreateGroupClickListener.onCreateGroupClicked(createPublicGroupName);
                }
            });
            holder.cardView.setOnLongClickListener(null);
            holder.cardView.setBackgroundResource(isSmall ? backgroundResId : R.drawable.clickable_green_4dp);
            return;
        }

        Group group = groups.get(position);

        holder.name.setText($(Val.class).of(group.getName(), $(ResourcesHandler.class).getResources().getString(R.string.app_name)));
        if (!group.hasEvent() && !group.isPublic()) {
            holder.cardView.setBackgroundResource(isSmall ? backgroundResId : R.drawable.clickable_blue_4dp);
            holder.action.setText(actionText != null ? actionText : $(ResourcesHandler.class).getResources().getString(R.string.open_group));
            holder.about.setText($(ResourcesHandler.class).getResources().getString(R.string.private_group));
        } else if (group.isPhysical()) {
            holder.cardView.setBackgroundResource(isSmall ? backgroundResId : R.drawable.clickable_purple_4dp);
            holder.action.setText(actionText != null ? actionText : $(ResourcesHandler.class).getResources().getString(R.string.open_group));
            holder.about.setText($(Val.class).of(group.getAbout()));
        } else if (group.hasEvent()) {
            holder.cardView.setBackgroundResource(isSmall ? backgroundResId : R.drawable.clickable_red_4dp);
            holder.action.setText(actionText != null ? actionText : $(ResourcesHandler.class).getResources().getString(R.string.open_event));
            Event event = $(StoreHandler.class).getStore().box(Event.class).query()
                    .equal(Event_.id, group.getEventId())
                    .build().findFirst();
            holder.about.setText(event != null ? $(EventDetailsHandler.class).formatEventDetails(event) :
                $(ResourcesHandler.class).getResources().getString(R.string.event));
        } else {
            holder.cardView.setBackgroundResource(isSmall ? backgroundResId : R.drawable.clickable_green_4dp);
            holder.action.setText(actionText != null ? actionText : $(ResourcesHandler.class).getResources().getString(R.string.open_group));
            holder.about.setText($(Val.class).of(group.getAbout()));
        }
        holder.cardView.setOnClickListener(view -> {
            if (onGroupClickListener != null) {
                onGroupClickListener.onGroupClicked(group, holder.itemView);
            }
        });
        holder.cardView.setOnLongClickListener(view -> {
            $(GroupMemberHandler.class).changeGroupSettings(group);
            return true;
        });

        holder.pool = tempPool();

        if (isSmall) {
            holder.actionRecyclerView.setVisibility(View.GONE);
        } else {
            holder.actionRecyclerView.setVisibility(View.VISIBLE);
            holder.pool.$(ApplicationHandler.class).setApp($(ApplicationHandler.class).getApp());
            holder.pool.$(ActivityHandler.class).setActivity($(ActivityHandler.class).getActivity());
            holder.pool.$(ApiHandler.class).setAuthorization($(AccountHandler.class).getPhone());
            holder.pool.$(GroupActionRecyclerViewHandler.class).attach(holder.actionRecyclerView, GroupActionAdapter.Layout.TEXT);
            holder.pool.$(GroupActionRecyclerViewHandler.class).setOnGroupActionRepliedListener(groupAction -> $(SearchHandler.class).openGroup(groupAction.getGroup(), holder.itemView));
            holder.pool.$(DisposableHandler.class).add($(StoreHandler.class).getStore().box(GroupAction.class).query()
                    .equal(GroupAction_.group, group.getId())
                    .build().subscribe().single()
                    .on(AndroidScheduler.mainThread())
                    .observer(groupActions -> {
                        holder.pool.$(GroupActionRecyclerViewHandler.class).getRecyclerView().setVisibility(groupActions.isEmpty() ? View.GONE : View.VISIBLE);
                        holder.pool.$(GroupActionRecyclerViewHandler.class).getAdapter().setGroupActions(groupActions);
                    }));

            $(ImageHandler.class).get().cancelRequest(holder.backgroundPhoto);
            if (group.getPhoto() != null) {
                holder.backgroundPhoto.setVisibility(View.VISIBLE);
                holder.backgroundPhoto.setImageDrawable(null);
                $(ImageHandler.class).get().load(group.getPhoto() + "?s=32")
                        .noPlaceholder()
                        .transform(new BlurTransformation($(ActivityHandler.class).getActivity(), 2))
                        .into(holder.backgroundPhoto);
            } else {
                holder.backgroundPhoto.setVisibility(View.GONE);
            }
        }

    }

    @Override
    public void onViewRecycled(@NonNull SearchGroupsViewHolder holder) {
        holder.pool.end();
    }

    @Override
    public int getItemCount() {
        return groups.size() + getCreateGroupCount();
    }

    public void setCreatePublicGroupName(String createPublicGroupName) {
        this.createPublicGroupName = createPublicGroupName;
        notifyDataSetChanged();
    }

    public void setGroups(List<Group> groups) {
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                return SearchGroupsAdapter.this.groups.size();
            }

            @Override
            public int getNewListSize() {
                return groups.size();
            }

            @Override
            public boolean areItemsTheSame(int oldPosition, int newPosition) {
                return SearchGroupsAdapter.this.groups.get(oldPosition).getId() != null &&
                        groups.get(newPosition).getId() != null &&
                        SearchGroupsAdapter.this.groups.get(oldPosition).getId().equals(groups.get(newPosition).getId());
            }

            @Override
            public boolean areContentsTheSame(int oldPosition, int newPosition) {
                return SearchGroupsAdapter.this.groups.get(oldPosition).getName() != null &&
                        groups.get(newPosition).getName() != null &&
                        SearchGroupsAdapter.this.groups.get(oldPosition).getName().equals(groups.get(newPosition).getName());
            }
        });
        this.groups.clear();
        this.groups.addAll(groups);
        diffResult.dispatchUpdatesTo(this);
    }

    private int getCreateGroupCount() {
        return (createPublicGroupName == null ? 0 : 1);
    }

    public void setActionText(String actionText) {
        this.actionText = actionText;
    }

    public void setIsSmall(boolean isSmall) {
        this.isSmall = isSmall;
    }

    protected class SearchGroupsViewHolder extends RecyclerView.ViewHolder {
        View cardView;
        TextView name;
        TextView about;
        TextView action;
        RecyclerView actionRecyclerView;
        TempPool pool;
        ImageView backgroundPhoto;

        public SearchGroupsViewHolder(View itemView) {
            super(itemView);
            cardView = itemView.findViewById(R.id.rootView);
            name = itemView.findViewById(R.id.name);
            about = itemView.findViewById(R.id.about);
            action = itemView.findViewById(R.id.action);
            actionRecyclerView = itemView.findViewById(R.id.actionRecyclerView);
            backgroundPhoto = itemView.findViewById(R.id.backgroundPhoto);
        }
    }

    public interface OnGroupClickListener {
        void onGroupClicked(Group group, View view);
    }

    public interface OnCreateGroupClickListener {
        void onCreateGroupClicked(String groupName);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/group/SearchGroupsAdapter.kt
package closer.vlllage.com.closer.handler.group

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.extensions.visible
import closer.vlllage.com.closer.handler.data.ApiHandler
import closer.vlllage.com.closer.handler.data.PersistenceHandler
import closer.vlllage.com.closer.handler.event.EventDetailsHandler
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.handler.quest.QuestHandler
import closer.vlllage.com.closer.pool.PoolRecyclerAdapter
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.*
import com.queatz.on.On
import io.objectbox.android.AndroidScheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.subjects.BehaviorSubject
import java.util.*

open class SearchGroupsAdapter constructor(
        on: On,
        private var showCreateGroup: Boolean,
        private val onGroupClickListener: ((group: Group, view: View) -> Unit)?,
        private val onCreateGroupClickListener: ((groupName: String, isPublic: Boolean) -> Unit)?
) : PoolRecyclerAdapter<RecyclerView.ViewHolder>(on) {

    private var createGroupName = BehaviorSubject.createDefault("")
    private var createIsPublic = true
    private val groups = mutableListOf<Group>()
    private var actionText: String? = null
    @LayoutRes
    private var layoutResId = R.layout.search_groups_item
    @DrawableRes
    private var backgroundResId = R.drawable.clickable_light
    private var isSmall: Boolean = false
    private var isNoAnimation: Boolean = false
    var flat: Boolean = false
    var transparentBackground: Boolean = false

    private val createGroupCount get() = if (showCreateGroup) 1 else 0

    fun setLayoutResId(layoutResId: Int): SearchGroupsAdapter {
        this.layoutResId = layoutResId
        return this
    }

    fun setBackgroundResId(backgroundResId: Int): SearchGroupsAdapter {
        this.backgroundResId = backgroundResId
        return this
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return SearchGroupsViewHolder(LayoutInflater.from(parent.context)
                .inflate(layoutResId, parent, false).also {
                    if (transparentBackground) {
                        it.background = null
                    }
                })
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder) {
            is SearchGroupsViewHolder -> {
                holder.on = On()

                if (position >= itemCount - createGroupCount) {
                    holder.action.text = on<ResourcesHandler>().resources.getString(R.string.continue_text)
                    holder.about.text = on<ResourcesHandler>().resources.getString(if (createIsPublic) R.string.add_new_public_group else R.string.add_new_private_group)
                    holder.backgroundPhoto.visible = false
                    setShadows(holder, false)
                    holder.actionRecyclerView.visible = false
                    holder.cardView.setOnClickListener {
                        onCreateGroupClickListener?.invoke(createGroupName.value
                                ?: "", createIsPublic)
                    }
                    holder.cardView.setOnLongClickListener(null)
                    holder.cardView.setBackgroundResource(if (isSmall) backgroundResId else if (createIsPublic) R.drawable.clickable_green_8dp else R.drawable.clickable_blue_8dp)

                    createGroupName.observeOn(AndroidSchedulers.mainThread()).subscribe {
                        holder.name.text = if (it.isNullOrBlank()) "+" else it
                    }.also { holder.on<DisposableHandler>().add(it) }
                    return
                }

                val group = groups[position]

                holder.cardView.setBackgroundResource(if (isSmall)
                    backgroundResId else on<GroupColorHandler>().getColorClickable8dp(group))

                holder.cardView.clipToOutline = true

                on<GroupNameHelper>().loadName(group, holder.name) { it }

                val recentActivity = (group.updated ?: Date(0)).after(on<TimeAgo>().weeksAgo())

                if (!group.hasEvent() && !group.isPublic && !group.physical) {
                    holder.action.text = if (actionText != null) actionText else if (recentActivity) on<TimeStr>().active(group.updated) else on<ResourcesHandler>().resources.getString(R.string.open_group)
                    holder.about.text = on<ResourcesHandler>().resources.getString(R.string.private_group)
                } else if (group.physical) {
                    holder.action.text = if (actionText != null) actionText else if (recentActivity) on<TimeStr>().active(group.updated) else on<ResourcesHandler>().resources.getString(R.string.open_group)
                    holder.about.text = group.about
                } else if (group.hasEvent()) {
                    holder.action.text = if (actionText != null) actionText else if (recentActivity) on<TimeStr>().active(group.updated) else on<ResourcesHandler>().resources.getString(R.string.open_event)
                    val event = on<StoreHandler>().store.box(Event::class).query()
                            .equal(Event_.id, group.eventId!!)
                            .build()
                            .findFirst()
                    holder.about.text = if (event != null)
                        on<EventDetailsHandler>().formatEventDetails(event)
                    else
                        on<ResourcesHandler>().resources.getString(R.string.event)
                } else if (group.ofKind == "quest") {
                    holder.action.text = on<ResourcesHandler>().resources.getString(R.string.open_quest)
                    holder.about.text = on<StoreHandler>().store.box(Quest::class).query()
                            .equal(Quest_.id, group.ofId!!)
                            .build()
                            .findFirst()?.let { quest -> on<QuestHandler>().questProgressText(on<StoreHandler>().store.box(QuestProgress::class).query()
                                    .equal(QuestProgress_.ofId, on<PersistenceHandler>().phoneId!!)
                                    .equal(QuestProgress_.questId, group.ofId!!)
                                    .equal(QuestProgress_.active, true)
                                    .build()
                                    .findFirst(), quest) }
                            ?: on<ResourcesHandler>().resources.getString(R.string.quest)
                } else {
                    holder.about.text = group.about
                    holder.action.text = if (actionText != null) actionText else if (recentActivity) on<TimeStr>().active(group.updated) else on<ResourcesHandler>().resources.getString(R.string.open_group)
                }

                if (flat) {
                    holder.container?.elevation = 0f
                }

                holder.cardView.setOnClickListener { view ->
                    onGroupClickListener?.invoke(group, holder.itemView)
                }
                holder.cardView.setOnLongClickListener { view ->
                    on<GroupMemberHandler>().changeGroupSettings(group)
                    true
                }

                if (isSmall) {
                    holder.actionRecyclerView.visible = false
                } else {
                    holder.actionRecyclerView.visible = true
                    holder.on.use(on<ApplicationHandler>())
                    holder.on.use(on<ActivityHandler>())
                    holder.on.use(on<ApiHandler>())
                    holder.on.use(on<StoreHandler>())

                    var questProgress: QuestProgress? = null
                    var quest: Quest? = null

                    if (group.ofKind == "quest") {
                        questProgress = on<StoreHandler>().store.box(QuestProgress::class).query()
                                .equal(QuestProgress_.ofId, on<PersistenceHandler>().phoneId!!)
                                .equal(QuestProgress_.questId, group.ofId!!)
                                .equal(QuestProgress_.active, true)
                                .build()
                                .findFirst()

                        quest = on<StoreHandler>().store.box(Quest::class).query()
                                .equal(Quest_.id, group.ofId!!)
                                .build()
                                .findFirst()
                    }

                    holder.on<GroupActionRecyclerViewHandler>().attach(holder.actionRecyclerView, GroupActionDisplay.Layout.TEXT, when {
                        quest != null -> ({ it, proceed ->
                            if (questProgress == null) {
                                holder.on<QuestHandler>().startQuest(quest) { questProgress ->
                                    if (questProgress != null) {
                                        holder.on<QuestHandler>().addProgress(quest, questProgress, it) { proceed() }
                                    } else {
                                        proceed()
                                    }
                                }
                            } else {
                                holder.on<QuestHandler>().addProgress(quest, questProgress, it) { proceed() }
                            }
                        })
                        else -> null
                    })
                    holder.on<DisposableHandler>().add(on<StoreHandler>().store.box(GroupAction::class).query().let {
                        if (group.ofKind == "quest") {
                            it.`in`(GroupAction_.id, quest?.flow?.items?.map { it.groupActionId!! }?.filter { groupActionId ->
                                        !on<QuestHandler>().isGroupActionProgressDone(quest, questProgress, groupActionId)
                                    }?.toTypedArray() ?: arrayOf())
                        } else {
                            it.equal(GroupAction_.group, group.id!!)
                        }
                    }
                            .build().subscribe().single()
                            .on(AndroidScheduler.mainThread())
                            .observer { groupActions ->
                                holder.on<GroupActionRecyclerViewHandler>().recyclerView!!.visible = groupActions.isNotEmpty()
                                holder.on<GroupActionRecyclerViewHandler>().adapter!!.setGroupActions(groupActions)
                            })

                    on<ImageHandler>().get().clear(holder.backgroundPhoto)
                    if (group.photo != null) {
                        holder.backgroundPhoto.visible = true
                        holder.backgroundPhoto.setImageDrawable(null)
                        on<PhotoLoader>().softLoad(group.photo!!, holder.backgroundPhoto)
                    } else {
                        holder.backgroundPhoto.visible = false
                    }

                    setShadows(holder, group.photo != null)
                }

                if (!isSmall) {
                    if (group.hasPhone() && group.photo == null) {
                        holder.on<LightDarkHandler>().setLight(true)
                    } else {
                        holder.on<LightDarkHandler>().setLight(false)
                    }

                    holder.on<LightDarkHandler>().onLightChanged.observeOn(AndroidSchedulers.mainThread()).subscribe {
                        holder.name.setTextColor(it.text)
                        holder.about.setTextColor(it.text)
                        holder.action.setTextColor(it.text)
                    }
                }
            }
        }
    }

    private fun setShadows(holder: SearchGroupsViewHolder, show: Boolean) {
        listOf(holder.name, holder.about, holder.action).forEach {
            it.setShadowLayer(
                    it.shadowRadius,
                    it.shadowDx,
                    it.shadowDy,
                    on<ResourcesHandler>().resources.getColor(if (show) R.color.black else R.color.black_transparent)
            )
        }
    }

    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        when (holder) {
            is SearchGroupsViewHolder -> {
                holder.on.off()
            }
        }
    }

    override fun getItemCount() = groups.size + createGroupCount

    fun setCreatePublicGroupName(createPublicGroupName: String) {
        this.createGroupName.onNext(createPublicGroupName)
    }

    fun setCreateIsPublic(isPublic: Boolean) {
        this.createIsPublic = isPublic
        notifyDataSetChanged()
    }

    fun showCreateOption(showCreate: Boolean) {
        if (showCreateGroup == showCreate) return
        showCreateGroup = showCreate
        notifyDataSetChanged()
    }

    fun setGroups(groups: List<Group>) {
        if (isNoAnimation || this.groups.isEmpty()) {
            this.groups.clear()
            this.groups.addAll(groups)
            notifyDataSetChanged()
        }

        val diffResult = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize() = this@SearchGroupsAdapter.groups.size

            override fun getNewListSize() = groups.size

            override fun areItemsTheSame(oldPosition: Int, newPosition: Int): Boolean {
                return this@SearchGroupsAdapter.groups[oldPosition].id == groups[newPosition].id
            }

            override fun areContentsTheSame(oldPosition: Int, newPosition: Int): Boolean {
                return this@SearchGroupsAdapter.groups[oldPosition].name == groups[newPosition].name &&
                        this@SearchGroupsAdapter.groups[oldPosition].photo == groups[newPosition].photo
            }
        })
        this.groups.clear()
        this.groups.addAll(groups)
        diffResult.dispatchUpdatesTo(this)
    }

    fun setActionText(actionText: String) {
        this.actionText = actionText
    }

    fun setIsSmall(isSmall: Boolean) {
        this.isSmall = isSmall
    }

    fun setNoAnimation(noAnimation: Boolean): SearchGroupsAdapter {
        isNoAnimation = noAnimation
        return this
    }

    class SearchGroupsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var container = itemView.findViewById<View>(R.id.container)
        var cardView: View = itemView.findViewById(R.id.rootView)
        var name: TextView = itemView.findViewById(R.id.name)
        var about: TextView = itemView.findViewById(R.id.about)
        var action: TextView = itemView.findViewById(R.id.action)
        var actionRecyclerView: RecyclerView = itemView.findViewById(R.id.actionRecyclerView)
        var backgroundPhoto: ImageView = itemView.findViewById(R.id.backgroundPhoto)
        lateinit var on: On
    }
}
